dpe/src/commands/mod.rs

```diff
@@ -67,26 +68,56 @@ impl Command {
     ///
     /// * `bytes` - serialized command
     pub fn deserialize(bytes: &[u8]) -> Result<Command, DpeErrorCode> {
-        let cmd_header = CommandHdr::try_from(bytes)?;
+        let header = CommandHdr::read_from_prefix(bytes).ok_or(DpeErrorCode::InvalidArgument)?;
+        if header.magic != CommandHdr::DPE_COMMAND_MAGIC {
+            return Err(DpeErrorCode::InvalidCommand);
+        }
+        // The client doesn't know what profile is implemented when calling the `GetProfile`
+        // command. But, all other commands should be directed towards the correct profile.
+        if header.cmd_id != Command::GET_PROFILE && header.profile != DPE_PROFILE as u32 {
+            return Err(DpeErrorCode::InvalidCommand);
+        }
+
         let bytes = &bytes[size_of::<CommandHdr>()..];
 
-        match cmd_header.cmd_id {
-            Command::GET_PROFILE => Ok(Command::GetProfile),
-            Command::INITIALIZE_CONTEXT => Ok(Command::InitCtx(InitCtxCmd::try_from(bytes)?)),
-            Command::DERIVE_CHILD => Err(DpeErrorCode::InvalidCommand),
-            Command::CERTIFY_KEY => Ok(Command::CertifyKey(CertifyKeyCmd::try_from(bytes)?)),
-            Command::SIGN => Ok(Command::Sign(SignCmd::try_from(bytes)?)),
-            Command::ROTATE_CONTEXT_HANDLE => {
-                Ok(Command::RotateCtx(RotateCtxCmd::try_from(bytes)?))
+        match header.cmd_id {
+            Self::GET_PROFILE
+            | Self::INITIALIZE_CONTEXT
+            | Self::DERIVE_CHILD
+            | Self::CERTIFY_KEY
+            | Self::SIGN
+            | Self::ROTATE_CONTEXT_HANDLE
+            | Self::DESTROY_CONTEXT
+            | Self::EXTEND_TCI
+            | Self::TAG_TCI
+            | Self::GET_TAGGED_TCI => {
+                Self::deserialize_args(&header, bytes).ok_or(DpeErrorCode::InvalidArgument)
             }
-            Command::DESTROY_CONTEXT => Ok(Command::DestroyCtx(DestroyCtxCmd::try_from(bytes)?)),
-            Command::GET_CERTIFICATE_CHAIN => Err(DpeErrorCode::InvalidCommand),
-            Command::EXTEND_TCI => Ok(Command::ExtendTci(ExtendTciCmd::try_from(bytes)?)),
-            Command::TAG_TCI => Ok(Command::TagTci(TagTciCmd::try_from(bytes)?)),
-            Command::GET_TAGGED_TCI => Ok(Command::GetTaggedTci(GetTaggedTciCmd::try_from(bytes)?)),
             _ => Err(DpeErrorCode::InvalidCommand),
         }
     }
+
+    fn deserialize_args(header: &CommandHdr, bytes: &[u8]) -> Option<Command> {
```

Comment:
I suspect you had to introduce this function because `read_from_prefix()` returns an `Option` and you wanted to factorize the `.ok_or(DpeErrorCode::InvalidArgument)`. But by doing so you also duplicate the list of supported commands. Another alternative would be to introduce a polymorphic helper function:

```rust
fn parse_command<T: FromBytes>(
    build: impl FnOnce(T) -> Command,
    bytes: &[u8],
) -> Result<Command, DpeErrorCode> {
    Ok(build(T::read_from_prefix(bytes).ok_or(DpeErrorCode::InvalidArgument)?))
}

match header.cmd_id {
    Command::GET_PROFILE => Ok(Command::GetProfile),
    Command::INITIALIZE_CONTEXT => parse_command(Command::InitCtx, bytes),
    ...
    _ => Err(DpeErrorCode::InvalidCommand),
}
```

I didn't test if it compile, but I suspect something along those lines to work.

Reply:
Great idea! That cleans it up a lot.